pkg/services/mssql/plan_details.go

```diff
@@ -0,0 +1,244 @@
+package mssql
+
+import (
+	"bytes"
+	"fmt"
+	"net"
+
+	"github.com/Azure/open-service-broker-azure/pkg/ptr"
+	"github.com/Azure/open-service-broker-azure/pkg/service"
+)
+
+const (
+	defaultCores            = 2
+	defaultVCoreStorageInGB = 5
+	maxStorageInGB          = 1024
+
+	gen5Hardware = "Gen5"
+)
+
+type planDetails interface {
+	getProvisionSchema() service.InputParametersSchema
+	getTierProvisionParameters(
+		service.Instance,
+	) (map[string]interface{}, error)
+}
+
+type dtuPlanDetails struct {
+	//sku        string
+	tier        string
+	skuMap      map[int64]string
+	allowedDTUs []int64
+	defaultDTUs int64
+	storageInGB int64
+	includeDBMS bool
+}
+
+func addDBMSParameters(schema map[string]service.PropertySchema) {
+	dbmsSchema := getDBMSCommonProvisionParamSchema().PropertySchemas
+	for key, value := range dbmsSchema {
+		schema[key] = value
+	}
+}
+
+func (d dtuPlanDetails) getProvisionSchema() service.InputParametersSchema {
+	schema := service.InputParametersSchema{
+		PropertySchemas: map[string]service.PropertySchema{},
+	}
+	if len(d.allowedDTUs) > 0 { //basic doesn't have DTUs, so don't add if not set
+		schema.PropertySchemas["dtu"] = service.IntPropertySchema{
+			AllowedValues: d.allowedDTUs,
+			DefaultValue:  ptr.ToInt64(d.defaultDTUs),
+			Description: "DTUs are a bundled measure of compute, " +
+				"storage, and IO resources.",
+		}
+	}
+	if d.includeDBMS {
+		addDBMSParameters(schema.PropertySchemas)
+	}
+	return schema
+}
+
+func (d dtuPlanDetails) getTierProvisionParameters(
+	instance service.Instance,
+) (map[string]interface{}, error) {
+	pp := databaseProvisionParams{}
+	if err := service.GetStructFromMap(
+		instance.ProvisioningParameters,
+		&pp,
+	); err != nil {
+		return nil, err
+	}
+	p := map[string]interface{}{}
+	p["sku"] = d.getSKU(pp)
+	p["tier"] = d.tier
+	p["maxSizeBytes"] = convertBytesToGB(d.storageInGB)
+	return p, nil
+}
+
+func (d dtuPlanDetails) getSKU(pp databaseProvisionParams) string {
+	if pp.DTUs != nil {
+		return d.skuMap[*pp.DTUs]
+	}
+	return d.skuMap[d.defaultDTUs]
+}
+
+type vCorePlanDetails struct {
+	tier          string
+	tierShortName string
+	includeDBMS   bool
+}
+
+func (v vCorePlanDetails) getProvisionSchema() service.InputParametersSchema {
+	schema := service.InputParametersSchema{
+		PropertySchemas: map[string]service.PropertySchema{
+			"cores": service.IntPropertySchema{
```

Comment:
I know this compiles, but is this working ok? (Schema is right and validations are as expected?) In the MySQL and PostgreSQL modules, I used references for all property schemas. _I can't remember why, but it might have had more to do with where I am headed with param access._ Looking at the functions on those types, the receivers all take non-references, so it would seem this _should_ be ok. (In the case of MySQL and PostgreSQL, Go is automatically dereferencing those references when the functions are invoked on them.) Bottom line-- no need to change this as long as you're certain it is working. If my upcoming PR breaks this, then that's on me to make the param access stuff more resilient so that this sensible bit of code here _doesn't_ break (and so module authors don't need to be hyper-aware of whether they should use a reference here or not).

Reply:
Good question! I just assumed it was working because it was compiling. Let me validate that it is.

Reply:
Validated.